### PR TITLE
feat: add folder

### PR DESCRIPTION
**Pull request by hacdias**
image

### CODE REVIEW

File: src/bundles/files.js [hunk 1]
```diff
@@ -119,26 +119,31 @@ function readAsBuffer (file) {
...
 
-  let promises = []
+  try {
+    let promises = []
 
-  for (const file of files) {
-    promises.push(readAsBuffer(file))
-  }
+    for (const file of files) {
+      promises.push(readAsBuffer(file))
```
Comment (lidel) [inline src/bundles/files.js]: This may be a separate PR, but we may want to replace readAsBuffer with pull stream to improve performance when adding big files:
- https://www.npmjs.com/package/filereader-pull-stream
- `files2stream()`: https://github.com/ipfs-shipyard/ipfs-companion/blob/b1bf771a37fbf0d6d11fab06317a8144a8c4962b/add-on/src/popup/quick-upload.js#L29-L41

File: src/bundles/files.js [hunk 2]
```diff
@@ -119,26 +119,31 @@ function readAsBuffer (file) {
...
+
+    for (const file of files) {
+      await getIpfs().files.mkdir(join(root, dirname(file.name)), { parents: true })
+    }
 
-  return Promise.all(promises).then(files => {
-    return Promise.all(files.map((file) => {
+    await Promise.all(files.map(file => {
       const target = join(root, file.name)
       return getIpfs().files.write(target, file.content, { create: true })
```
Comment (lidel) [inline src/bundles/files.js]: We should  avoid `files.write`, as it will produce different CID than `add`.
See details and workaround in: https://github.com/ipfs-shipyard/ipfs-webui/issues/676

**Comment by hacdias** [comment]
@lidel I just made the files.write change and gave the dropdown some styling. I'll add each feature in a different PR to keep everything small. Could you see if the code's looking good?

File: src/bundles/files.js [hunk 3]
```diff
@@ -119,26 +119,35 @@ function readAsBuffer (file) {
...
+      const path = join(root, dirname(file.name))
+      await getIpfs().files.mkdir(path, { parents: true })
+    }
 
-  return Promise.all(promises).then(files => {
-    return Promise.all(files.map((file) => {
-      const target = join(root, file.name)
-      return getIpfs().files.write(target, file.content, { create: true })
+    await Promise.all(files.map(async file => {
+      const res = await getIpfs().add([file.content])
```
Comment (lidel) [inline src/bundles/files.js]: I wonder if we should disable pinning when `add`ing for MFS. 
As long as files are in MFS, they won't be GC'd.
If we dont pass `{pin: false}` to `add` then removing file from MFS won't remove it from node.  
Is that a [bug or feature](https://ipfs.io/ipfs/QmVnLkRxWWv2xiPKeQy6re1zCepzX7YXUiq5fESkd1j8bC/bug_or_feature.jpg) ?
@olizilla  any thoughts on this?

Comment (hacdias) [reply src/bundles/files.js]: I'm adding that option then.

About the progress of uploading: don't worry. Won't be doing in this PR but it is on Invision how it is supposed to be.

Comment (hacdias) [reply src/bundles/files.js]: Done @lidel

**Comment by hacdias** [comment]
Done @lidel

File: src/bundles/files.js (hunk 1 above)
Comment (olizilla) [reply src/bundles/files.js]: Yes please. Is ok for separate PR, but we've gotta make this the best file uploader we can.

File: src/lib/path.js [hunk 4]
```diff
@@ -104,3 +104,9 @@ export function join (a, b) {
   if (b.startsWith('/')) b = b.slice(1)
   return `${a}/${b}`
 }
+
+export function dirname (path) {
```
Comment (olizilla) [inline src/lib/path.js]: Let's just use node core `path` stuff for this kind of thing, https://nodejs.org/api/path.html#path_path_dirname_path

Comment (olizilla) [reply src/lib/path.js]: Of note this file was intended to be a lib for ipfs/ipld path helpers, rather than any sort of path.

**Comment by hacdias** [comment]
Done @olizilla

**Comment by olizilla** [comment]
if nothing else is using our `join` impl, can you remove it pls. If there is usages, pls update them.

**Comment by hacdias** [comment]
Done @olizilla. Since I've already solved the issues you and @lidel mentioned, I'm going to merge this so I can continue to the next feature 😄
